alns/stop/tests/test_no_improvement.py

```diff
@@ -39,40 +52,42 @@ def test_first_iteration():
     stop = NoImprovement(100)
     rnd = RandomState()
 
-    assert not stop(rnd, Zero(), Zero())
+    assert_(not stop(rnd, Zero(), One()))
 
 
-def test_before_n_iterations():
+def test_stop_after_max_iterations():
     stop = NoImprovement(100)
     rnd = RandomState()
 
     for _ in range(100):
         assert_(not stop(rnd, Zero(), Zero()))
 
+    for _ in range(100):
+        assert_(stop(rnd, Zero(), Zero()))
 
-def test_after_n_iterations():
+
+def test_counter_value_at_stop():
+    """
+    Test if the counter value is equal to the passed-in max iterations
+    at the first stopping occurence.
+    """
     stop = NoImprovement(100)
-    rnd = RandomState()
 
-    for _ in range(100 + 1):
-        stop(rnd, Zero(), Zero())
+    while not stop(RandomState(), Zero(), Zero()):
+        pass
 
-    for _ in range(100):
-        # The 102-th iteration and beyond should be stopped, because there was
-        # 1 iteration to initialize and 100 subsequent non-improving iterations.
-        assert_(stop(rnd, Zero(), Zero()))
+    assert_equal(stop._counter, 100)
 
 
-@pytest.mark.parametrize("n_iterations", [10, 100, 1000])
-def test_reset_counter(n_iterations):
+def test_reset_counter():
     """
-    Test if the counter is reset when an improving solution is encountered.
+    Test if the counter is reset when a new best solution is encountered.
     """
-    stop = NoImprovement(n_iterations)
+    stop = NoImprovement(100)
     rnd = RandomState()
 
-    for _ in range(n_iterations):
-        assert_(not stop(rnd, Two(), Zero()))
+    for _ in range(5):
+        assert_(not stop(rnd, One(), One()))
 
-    assert not stop(rnd, One(), Zero())
-    assert stop._counter == 0
+    assert_(not stop(rnd, Zero(), One()))
+    assert_equal(stop._counter, 0)
```

Comment:
Refactor: we should not care if the `NoImprovement` class implements its behaviour using a counter or not: we only care that after a given number of iterations without improvement, it provides a stop signal (and not in all other occasions).

I suggest we remove this final test.

Reply:
Following the same principle, I also removed `test_counter_value_at_stop`.